pkg/river/internal/value/decode.go

```diff
@@ -0,0 +1,385 @@
+package value
+
+import (
+	"encoding"
+	"fmt"
+	"reflect"
+)
+
+// Decode assigns a Value val to a Go pointer target. Decode will attempt to
+// convert val to the type expected by target for assignment. If val cannot be
+// converted, an error is returned. Pointers will be allocated as necessary
+// when decoding.
+//
+// New arrays and slices will be allocated when decoding into target.
+//
+// River null values will decode into a nil Go pointer or the zero value for
+// the non-pointer type.
+//
+// Decode will panic if target is not a pointer.
+func Decode(val Value, target interface{}) error {
+	rt := reflect.ValueOf(target)
+	if rt.Kind() != reflect.Pointer {
+		panic("river/value: Decode called with non-pointer value")
+	}
+	return decode(val, rt)
+}
+
+func decode(val Value, into reflect.Value) error {
+	// Before decoding, we need to temporarily take the address of rt so we can
+	// handle the case of it implementing supported interfaces.
+	if into.CanAddr() {
+		into = into.Addr()
+	}
+
+	switch {
+	case into.Type().Implements(goTextUnmarshaler):
+		var s string
+		err := decode(val, reflect.ValueOf(&s))
+		if err != nil {
+			return err
+		}
+		err = into.Interface().(encoding.TextUnmarshaler).UnmarshalText([]byte(s))
+		if err != nil {
+			return Error{Value: val, Inner: err}
+		}
+		return nil
+	}
+
+	// Fully deference rt and allocate pointers as necessary.
+	for into.Kind() == reflect.Pointer {
+		// If our value is Null, we want to set the first settable pointer to nil.
+		if val.Type() == TypeNull && into.CanSet() {
+			into.Set(reflect.Zero(into.Type()))
+			return nil
+		}
+
+		if into.IsNil() {
+			into.Set(reflect.New(into.Type().Elem()))
+		}
+		into = into.Elem()
+	}
+
+	// Fastest cases: we can directly assign values.
+	switch {
+	case val.Type() == TypeNull:
+		// TODO(rfratto): Does it make sense for a null to always decode into the
+		// zero value? Maybe only objects and arrays should support null?
+		into.Set(reflect.Zero(into.Type()))
```

Comment:
This feels like something we may pivot on has we get experience but seems reasonable at the moment. I wonder if we could get away with a null-less language

Reply:
Might be possible, the challenges would be figuring out: 

1. What does a Go nil value encode to? Does it fail to encode?
2. If we have a decode_json function that includes a null, what does that convert to?

We can keep it in the back of our minds as we build this out, we might get some ideas of what we can do